File: python/189.轮转数组.py
#
# @lc app=leetcode.cn id=189 lang=python3
#
# [189] 轮转数组
#
from typing import List
import math

# @lc code=start
class Solution:
    # 环形替换， 用到了数学知识
    def rotate(self, nums: List[int], k: int) -> None:
        n = len(nums)
        k = k % n
        gcd = math.gcd(n, k)
        for start in range(0, gcd):
            cur = start
            pre = nums[start]
            while True:
                next = (cur + k) % n
                nums[next], pre = pre, nums[next]
                cur = next
                if cur == start:
                    break

    # 数组翻转
    def rotate3(self, nums: List[int], k: int) -> None:
        n = len(nums)
        offset = k % n
        self.swap(nums, 0, n - 1)
        self.swap(nums, 0, offset - 1)
        self.swap(nums, offset, n - 1)

    # 复制数组，然后赋值(通用)
    def rotate2(self, nums: List[int], k: int) -> None:
        n = len(nums)
        tempArr = [0] * n
        for i, _ in enumerate(nums):
            tempArr[(i + k) % n] = nums[i]
        nums[:] = tempArr[:]

    # 复制数组，然后赋值(python 语法)
    def rotate1(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        offset = k % len(nums)
        if not offset:
            return
        copy = nums[:]
        nums[:offset] = copy[-offset:]
        nums[offset:] = copy[:-offset]

    def swap(self, nums: List[int], left: int, right: int) -> None:
        while left < right:
            temp = nums[left]
            nums[left] = nums[right]
            nums[right] = temp
            left += 1
            right -= 1

# Solution().rotate([1], 0)
# @lc code=end

File: cpp/189.轮转数组.cpp
/*
 * @lc app=leetcode.cn id=189 lang=cpp
 *
 * [189] 轮转数组
 */
#include <list>
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    void rotate(vector<int>& nums, int k)
    {
        return method3(nums, k);
    }

    // 翻转数组
    void method3(vector<int>& nums, int k) {
        int n = nums.size();
        k = k % n;
        reverse(nums, 0, n - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, n - 1);
    }

    void reverse(vector<int>& nums, int left, int right)
    {
        while (left < right) {
            swap(nums[left++], nums[right--]);
        }
    }

    // 稍微优雅一点的恐惧O(n)写法
    void method2(vector<int>& nums, int k)
    {
        int n = nums.size();
        vector<int> tmp(n, n);
        for (int i = 0; i < n; i++) {
            tmp[(i + k) % n] = nums[i];
        }
        nums.assign(tmp.begin(), tmp.end());
    }

    // 丑陋的直观写法
    void method1(vector<int>& nums, int k)
    {
        list<int> S;

        int n = nums.size();
        k     = k % n;

        for (auto i : nums) {
            S.push_back(i);
        }

        for (int i = k; i > 0; i--) {
            nums[i - 1] = S.back();
            S.pop_back();
        }

        for (int i = k; i < n; i++) {
            nums[i] = S.front();
            S.pop_front();
        }
    }
};
// @lc code=end
